tao_compiler/mlir/disc/transforms/disc_cse.cc

```diff
@@ -0,0 +1,219 @@
+    if (lhs == rhs) return true;
+
+    // Compare the operation properties.
+    if (lhs->getName() != rhs->getName() ||
+        lhs->getAttrDictionary() != rhs->getAttrDictionary() ||
+        lhs->getNumRegions() != rhs->getNumRegions() ||
+        lhs->getNumSuccessors() != rhs->getNumSuccessors() ||
+        lhs->getNumOperands() != rhs->getNumOperands() ||
+        lhs->getNumResults() != rhs->getNumResults())
+      return false;
+    if (!(flags & OperationEquivalence::IgnoreLocations) &&
+        lhs->getLoc() != rhs->getLoc())
+      return false;
+
+    ValueRange lhsOperands = lhs->getOperands(),
+               rhsOperands = rhs->getOperands();
+    SmallVector<Value> lhsOperandStorage, rhsOperandStorage;
+    if (lhs->hasTrait<mlir::OpTrait::IsCommutative>()) {
+      lhsOperandStorage.append(lhsOperands.begin(), lhsOperands.end());
+      llvm::sort(lhsOperandStorage, [](Value a, Value b) -> bool {
+        return a.getAsOpaquePointer() < b.getAsOpaquePointer();
+      });
+      lhsOperands = lhsOperandStorage;
+
+      rhsOperandStorage.append(rhsOperands.begin(), rhsOperands.end());
+      llvm::sort(rhsOperandStorage, [](Value a, Value b) -> bool {
+        return a.getAsOpaquePointer() < b.getAsOpaquePointer();
+      });
+      rhsOperands = rhsOperandStorage;
+    }
+    auto checkValueRangeMapping =
+        [](ValueRange lhs, ValueRange rhs,
+           function_ref<LogicalResult(Value, Value)> mapValues) {
+          for (auto operandPair : llvm::zip(lhs, rhs)) {
+            Value curArg = std::get<0>(operandPair);
+            Value otherArg = std::get<1>(operandPair);
+            if (curArg.getType() != otherArg.getType()) return false;
+            if (failed(mapValues(curArg, otherArg))) return false;
+          }
+          return true;
+        };
+    // Check mapping of operands and results.
+    if (!checkValueRangeMapping(lhsOperands, rhsOperands, mapOperands))
+      return false;
+    if (!checkValueRangeMapping(lhs->getResults(), rhs->getResults(),
+                                mapResults))
+      return false;
+
+    auto lhsReduceOp = dyn_cast<mhlo::ReduceOp>(lhs);
+    auto rhsReduceOp = dyn_cast<mhlo::ReduceOp>(rhs);
+    if (lhsReduceOp && rhsReduceOp) {
+      auto& lhsBody = lhsReduceOp.getBody();
+      auto& rhsBody = rhsReduceOp.getBody();
+      if (not(lhsBody.hasOneBlock() && rhsBody.hasOneBlock())) return false;
+
+      for (auto it : llvm::zip(lhsBody.front().getOperations(),
+                               rhsBody.front().getOperations())) {
+        auto& lhsOp = std::get<0>(it);
+        auto& rhsOp = std::get<1>(it);
+        // We currently do not compare equality of nested regions
+        if (lhsOp.getNumRegions() != 0 or rhsOp.getNumRegions() != 0)
+          return false;
+
+        auto innerOpMatch = OperationEquivalence::isEquivalentTo(
+            &lhsOp, &rhsOp,
+            /*mapOperands=*/OperationEquivalence::ignoreValueEquivalence,
```

Comment:
I think it's not safe to ignore operand value check here.

For example:
```

mhlo.reduce(...) {
^bb0(%arg0 : f32, %arg1 : f32) {
   %0 = arith.mulf %arg0, %c2 // const is 2
   %1 = arith.muf %0, %arg1
   return %1 : f32 
 }

mhlo.reduce(...) {
^bb0(%arg0 : f32, %arg1 : f32) {
   %0 = arith.mulf %arg0, %c3 // here is const 3
   %1 = arith.muf %0, %arg1
   return %1 : f32 
 }
```

maybe we can support more restricted reduce op, for example those only have one known op (e.g. add/sub/mul/div) in the body.

Reply:
Indeed, `OperationEquivalence::isEquivalentTo` will compare op properties first. Thus different Constant values will not be regarded as equivalent. See https://mlir.llvm.org/doxygen/OperationSupport_8cpp_source.html#L712

Also, I added a test to the case you mentioned above, see `tests/disc_mlir/disc_cse.mlr:multiple_reduce_with_cst0`

Reply:
constant op may not defined inside the region of reduce op (in my original example above). Is it still ok in such case?

```
    %3 = mhlo.constant dense<1.000000e+00> : tensor<f16>
    %4 = mhlo.reduce(%arg0 init: %0) across dimensions = [0, 2, 3] : (tensor<?x?x?x?xf16>, tensor<f16>) -> tensor<?xf16>
     reducer(%arg1: tensor<f16>, %arg2: tensor<f16>)  {
      %6 = mhlo.multiply %arg1, %3 : tensor<f16>
      %7 = mhlo.add %6, %arg2 : tensor<f16>
      mhlo.return %7 : tensor<f16>
    }
```